design_documents/parametrizing_tests_with_fixed_values.md

```diff
@@ -0,0 +1,205 @@
+# Parametrizing Tests With Fixed Values
+
+## Context
+
+Various test runners for other programming languages provide ways of parametrizing test cases with fixed values.
+This way, multiple test cases can be executed for the same test source code.
+
+## Goal
+
+Propose a way of adding support for parametrized tests to Starknet Foundry's `snforge`.
+
+## Proposed Solution
+
+Introduce new attribute `test_case`.
+
+The attribute should accept arguments:
+
+- Optional, named argument `name`.
+- Unnamed, non-optional arguments, each corresponding to a test argument.
+
+Generate code for each of the test cases so no handling of values injection, etc. is necessary.
+The compiler would then validate if the values for the generated code are actually valid.
+Thanks to that, we could have failures for invalid arguments early in the execution.
+
+### Parametrization With Arguments
+
+In the `test_case` attribute, user provides a list of unnamed arguments, each corresponding to test function argument.
+Arguments are handled in order: The first argument of the attribute corresponds to the first argument of a test
+function.
+
+```cairo
+#[test_case(1, 3)]
+#[test_case(3, 5)]
+fn my_test(a: felt252, b: u32) {
+    // ...
+}
+```
+
+For test cases like this, we would generate two actual test cases:
+
+```cairo
+#[test]
+fn my_test_case_1() {
```

Comment:
How will named test cases be generated? They shouldn't break unnamed test cases, if they are named "1" for example

Reply:
I think it's an implementation detail

Reply:
I've considered adding some kind of a prefix to the test name to mark it's user generated, but that's for an implementer to handle.